libcudacxx/include/cuda/std/__type_traits/type_set.h

```diff
@@ -21,18 +21,29 @@
 #endif // no system header
 
 #include <cuda/std/__type_traits/conditional.h>
+#include <cuda/std/__type_traits/conjunction.h>
 #include <cuda/std/__type_traits/fold.h>
+#include <cuda/std/__type_traits/integral_constant.h>
 #include <cuda/std/__type_traits/is_base_of.h>
 #include <cuda/std/__type_traits/type_identity.h>
 #include <cuda/std/cstddef>
 
-#if _CCCL_STD_VER >= 2014
-
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
+#ifndef _CCCL_NO_VARIABLE_TEMPLATES
+
+template <class _Set, class... _Ty>
+_CCCL_INLINE_VAR constexpr bool __type_set_contains_v = __fold_and_v<is_base_of_v<__type_identity<_Ty>, _Set>...>;
+
+template <class _Set, class... _Ty>
+using __type_set_contains = bool_constant<__type_set_contains_v<_Set, _Ty...>>;
+
+#else // ^^^ !_CCCL_NO_VARIABLE_TEMPLATES ^^^ / vvv _CCCL_NO_VARIABLE_TEMPLATES vvv
+
 template <class _Set, class... _Ty>
-_CCCL_INLINE_VAR constexpr bool __type_set_contains =
-  _CUDA_VSTD::__fold_and<_CCCL_TRAIT(_CUDA_VSTD::is_base_of, __type_identity<_Ty>, _Set)...>;
+using __type_set_contains = __fold_and<is_base_of<__type_identity<_Ty>, _Set>::value...>;
+
+#endif // _CCCL_NO_VARIABLE_TEMPLATES
```

Comment:
Is there a benefit to special casing __type_set_contains instead of just always using `__fold_and` ?

Reply:
you would need to use a nested `__fold_and` to test whether each type in the pack is a member of the set, making the algorithm O(n^2). by building a set from the types, we can query set membership in O(1), so testing a pack for set membership is O(n).

Reply:
Oh I believe we mean different things. I meant:

```suggestion
template <class _Set, class... _Ty>
using __type_set_contains = __fold_and<is_base_of<__type_identity<_Ty>, _Set>::value...>;

#ifndef _CCCL_NO_VARIABLE_TEMPLATES
template <class _Set, class... _Ty>
_CCCL_INLINE_VAR constexpr bool __type_set_contains_v = __fold_and_v<is_base_of_v<__type_identity<_Ty>, _Set>...>;
#endif //_CCCL_NO_VARIABLE_TEMPLATES
```

Reply:
i want `__type_set_contains<S,Ts...>` to be lazy and do no work until a `::value` is requested. see how `__type_set_eq` is using it. in general, when a trait has both a type and a variable template form, i want to type form to be lazy.